### PR TITLE
Fix overlapping samples in DDP when no global seed is set

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

This fixes an edge case when the user wants to run with `shuffle=True`, no global seed set, accelerator="ddp". On master, this situation results in overlapping data. Reported by user Turab Iqbal on Slack (thanks!).

Repro:

```py
import lightning as pl
import torch
​
​
class MyModule(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(1, 1)
​
    def configure_optimizers(self):
        return torch.optim.SGD(self.parameters(), lr=0.1)
​
    def training_step(self, batch, _):
        print(f'value: {batch.item()}')
​
​
loader = torch.utils.data.DataLoader(range(16), shuffle=True)
trainer = pl.Trainer(
    accelerator='gpu',
    devices=2,
    strategy='ddp_find_unused_parameters_false',
    enable_progress_bar=False,
    max_epochs=1,
    logger=False,
)
trainer.fit(MyModule(), loader)
```

Note:
- This is only a problem with DDP proper. DDP-spawn and fork don't have this issue since they copy the program state and start with the same initial state on all ranks. 
- This is not an issue for Fabric, and the changes here align the logic for both implementations.

cc @borda @justusschock @awaelchli

### CODE REVIEW

**Review by justusschock — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: tests/tests_pytorch/trainer/test_dataloaders.py [hunk 1]
```diff
@@ -804,6 +804,40 @@ def test_dataloader_distributed_sampler(tmpdir):
...
+    """Test that the samples are non-overlapping in DDP when shuffling is enabled and no global seed is set."""
+    # This test must run without a global seed set (e.g. through `seed_everything`), to ensure that each process
+    # starts with a different initial state.
+    assert "PL_GLOBAL_SEED" not in os.environ
+    train_dataloader = DataLoader(range(32), shuffle=True, batch_size=4)
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        num_sanity_val_steps=False,
+        logger=False,
+        enable_progress_bar=False,
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/test_dataloaders.py:832]:
```suggestion
        barebones=True,
```

File: src/lightning/pytorch/trainer/connectors/data_connector.py [hunk 2]
```diff
@@ -251,8 +251,11 @@ def _get_distributed_sampler(
     """This function is used to created the distributed sampler injected within the user DataLoader."""
     kwargs["shuffle"] = shuffle and not overfit_batches
     kwargs.setdefault("seed", int(os.getenv("PL_GLOBAL_SEED", 0)))
-    cls = UnrepeatedDistributedSamplerWrapper if mode == RunningStage.PREDICTING else DistributedSamplerWrapper
-    return cls(dataloader.sampler, **kwargs)
+    if mode == RunningStage.PREDICTING:
+        return UnrepeatedDistributedSamplerWrapper(dataloader.sampler, **kwargs)
+    if isinstance(dataloader.sampler, (RandomSampler, SequentialSampler)):
```
Comment (carmocca) [inline src/lightning/pytorch/trainer/connectors/data_connector.py:256]: I would suggest that this `if` is put in the `DistributedSamplerWrapper` __init__.

It's docstring advertises it as

> Allows you to use any sampler in distributed mode. It will be automatically used by Lightning in distributed mode if sampler replacement is enabled.

Putting it here invalidates that statement.

Comment (awaelchli) [reply src/lightning/pytorch/trainer/connectors/data_connector.py:256]: The `DistributedSamplerWrapper` is a confusing class. Putting this condition into its init won't be enough. Its implementation would have to be changed. It is much more sane to put this logic here, because the PyTorch Distributed sampler supports handling these two.

Comment (carmocca) [reply src/lightning/pytorch/trainer/connectors/data_connector.py:256]: My point is that that class is meant to encapsulate this kind of logic. The `DistributedSamplerWrapper` is itself a `DistributedSampler`. If one was using it standalone (possible, as it's public API), then the reported issue should be fixed for it too.

I think the ideal fix would look into why it has this problem, but the if-else workaround proposed here is good to quickly fix the issue. The implementation changes would be minimal as you could check for `isinstance(_DatasetSamplerWrapper)` to split the logic

Comment (awaelchli) [reply src/lightning/pytorch/trainer/connectors/data_connector.py:256]: > I think the ideal fix would look into why it has this problem

It is obvious why this is the case. The custom-distributed sampler is bypassing the entire seeding logic in the original implementation, which is not a problem if you set a global seed. I disagree with your proposal, there is no point in using the custom logic for the standard samplers in PyTorch. Can you open a PR yourself to redesign the custom class to your liking?

**Review by carmocca — APPROVED** [review]

File: src/lightning/pytorch/trainer/connectors/data_connector.py (hunk 2 above)
Comment (carmocca) [reply src/lightning/pytorch/trainer/connectors/data_connector.py:256]: Approved to unblock
